### PR TITLE
Weekly tx report for zrl

### PR DESCRIPTION
**Pull request by malted**

### CODE REVIEW

File: app/mailers/transaction_report_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+      cpts = CanonicalPendingTransaction.not_declined.joins(:raw_pending_stripe_transaction).where("(stripe_transaction->'card'->>'id' IN (?)) AND (CAST(stripe_transaction->>'created' AS BIGINT) >= EXTRACT(EPOCH FROM TIMESTAMP ?))", card_ids, 1.week.ago)
+      cts = CanonicalTransaction.stripe_transaction.where("(stripe_transaction->'card'->>'id' IN (?)) AND (CAST(stripe_transaction->>'created' AS BIGINT) >= EXTRACT(EPOCH FROM TIMESTAMP ?))", card_ids, 1.week.ago)
+
+      { user:, txs: cpts + cts }
+    end.compact
+
+    @total = @target_users.map{ |x| x[:txs].pluck(:amount_cents).sum }.sum * -1
+
+    mail to: "zach@hackclub.com", subject: "Your weekly transaction report", from: "hcb@hackclub.com"
```
Comment (cjdenio) [inline app/mailers/transaction_report_mailer.rb]:
```suggestion
    mail to: "zach@hackclub.com", subject: "Your weekly transaction report"
```

File: app/mailers/transaction_report_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class TransactionReportMailer < ApplicationMailer
+  def tell_zach
+    target_user_emails = ["thomas@hackclub.com", "deet@hackclub.com", "amanda@hackclub.com", "dev@hackclub.com"]
+
+    @target_users = target_user_emails.map do |email|
+      user = User.find_by(email:)
+      next unless user
```
Comment (cjdenio) [inline app/mailers/transaction_report_mailer.rb]:
```suggestion
    target_users = User.where email: ["thomas@hackclub.com", "deet@hackclub.com", "amanda@hackclub.com", "dev@hackclub.com"]

    @target_users = target_users.map do |user|
```

a little more efficient

File: app/mailers/transaction_report_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+      card_ids = user.stripe_cards.filter { |card| card.event.category != "salary" }.pluck(:stripe_id)
+
+      cpts = CanonicalPendingTransaction.not_declined.joins(:raw_pending_stripe_transaction).where("(stripe_transaction->'card'->>'id' IN (?)) AND (CAST(stripe_transaction->>'created' AS BIGINT) >= EXTRACT(EPOCH FROM TIMESTAMP ?))", card_ids, 1.week.ago)
```
Comment (cjdenio) [inline app/mailers/transaction_report_mailer.rb]:
```suggestion
      cpts = CanonicalPendingTransaction.unsettled.joins(:raw_pending_stripe_transaction).where("(stripe_transaction->'card'->>'id' IN (?)) AND (CAST(stripe_transaction->>'created' AS BIGINT) >= EXTRACT(EPOCH FROM TIMESTAMP ?))", card_ids, 1.week.ago)
```
needed so settled transactions don't show up twice

File: app/mailers/transaction_report_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      cts = CanonicalTransaction.stripe_transaction.where("(stripe_transaction->'card'->>'id' IN (?)) AND (CAST(stripe_transaction->>'created' AS BIGINT) >= EXTRACT(EPOCH FROM TIMESTAMP ?))", card_ids, 1.week.ago)
+
+      { user:, txs: cpts + cts }
+    end.compact
+
+    @total = @target_users.map{ |x| x[:txs].pluck(:amount_cents).sum }.sum * -1
```
Comment (cjdenio) [inline app/mailers/transaction_report_mailer.rb]:
```suggestion
    @total = @target_users.map{ |x| x[:txs].sum(&:amount_cents)}.sum.abs
```

File: app/views/transaction_report_mailer/tell_zach.html.erb [hunk 5]
```diff
@@ -0,0 +1,63 @@
+<p>Salutations Zachary.</p>
+
+<p>Your reports spent <%= render_money @total %> in total this week. Here's a breakdown;</p>
+
+<% @target_users.each do |user_info| %>
+    <h2><%= user_info[:user].full_name %></h2>
+    <p><%= user_info[:user].name %> spent <%= render_money(user_info[:txs].pluck(:amount_cents).sum * -1) %> this week</p>
```
Comment (cjdenio) [inline app/views/transaction_report_mailer/tell_zach.html.erb]:
```suggestion
    <p><%= user_info[:user].name %> spent <%= render_money(user_info[:txs].pluck(:amount_cents).sum.abs) %> this week</p>
```

File: app/mailers/transaction_report_mailer.rb (hunk 4 above)
Comment (malted) [reply app/mailers/transaction_report_mailer.rb]: Won't `abs` make refunds look the same (have the same sign) as charges?

File: app/views/transaction_report_mailer/tell_zach.html.erb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      <tr>
+        <th>Memo</th>
+        <th>Amount</th>
+        <th>Event</th>
+        <th>Receipts</th>
+      </tr>
+      <% user_info[:txs].sort_by(&:amount_cents).each do |tx| %>
+          <tr>
+            <td>
+              <a href="<%= tx.local_hcb_code.url %>">
```
Comment (garyhtou) [inline app/views/transaction_report_mailer/tell_zach.html.erb]: This should be an absolute link

Comment (garyhtou) [reply app/views/transaction_report_mailer/tell_zach.html.erb]: `url_for(tx.local_hcb_code)`

if `url_for` is undefined, then maybe try `Rails.application.routes.url_helpers.url_for`

File: app/views/transaction_report_mailer/tell_zach.html.erb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                <%= tx.memo %>
```
Comment (garyhtou) [inline app/views/transaction_report_mailer/tell_zach.html.erb]: tx.local_hcb_code.memo

File: app/views/transaction_report_mailer/tell_zach.html.erb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+              </a>
+            </td>
+            <td>
+              <%= render_money tx.amount_cents * -1 %>
+            </td>
+            <td>
+                <a href="/<%= tx.event.slug %>">
```
Comment (garyhtou) [inline app/views/transaction_report_mailer/tell_zach.html.erb]: needs to be an absolute path

**Pull request merged**